File: src/topo_shadow_box/core/gpx.py
"""GPX file parsing."""

import gpxpy
from topo_shadow_box.models import GpxTrack, GpxPoint, GpxWaypoint


def parse_gpx_file(filepath: str) -> dict:
    """Parse a GPX file and extract tracks, waypoints, and bounds."""
    with open(filepath, "r") as f:
        gpx = gpxpy.parse(f)

    tracks = []
    for track in gpx.tracks:
        points = []
        for segment in track.segments:
            for point in segment.points:
                points.append(GpxPoint(
                    lat=point.latitude,
                    lon=point.longitude,
                    elevation=point.elevation if point.elevation else 0.0,
                ))
        if len(points) >= 2:
            tracks.append(GpxTrack(
                name=track.name or "Unnamed Track",
                points=points,
            ))

    waypoints = []
    for wp in gpx.waypoints:
        waypoints.append(GpxWaypoint(
            name=wp.name or "",
            lat=wp.latitude,
            lon=wp.longitude,
            elevation=wp.elevation if wp.elevation else 0.0,
            description=wp.description or "",
        ))

    bounds = gpx.get_bounds()
    bounds_dict = None
    if bounds:
        bounds_dict = {
            "north": bounds.max_latitude,
            "south": bounds.min_latitude,
            "east": bounds.max_longitude,
            "west": bounds.min_longitude,
        }

    return {
        "tracks": tracks,        # now list[GpxTrack]
        "waypoints": waypoints,  # now list[GpxWaypoint]
        "bounds": bounds_dict,
        "metadata": {
            "name": gpx.name,
            "description": gpx.description,
        },
    }

File: go/internal/core/gpx.go
package core

import (
	"encoding/xml"
	"fmt"
	"io"
	"os"

	"github.com/huslage/topo-shadow-box/internal/session"
)

type gpxDoc struct {
	Tracks    []gpxTrack    `xml:"trk"`
	Waypoints []gpxWaypoint `xml:"wpt"`
}

type gpxTrack struct {
	Name     string       `xml:"name"`
	Segments []gpxSegment `xml:"trkseg"`
}

type gpxSegment struct {
	Points []gpxTrackPoint `xml:"trkpt"`
}

type gpxTrackPoint struct {
	Lat float64  `xml:"lat,attr"`
	Lon float64  `xml:"lon,attr"`
	Ele *float64 `xml:"ele"`
}

type gpxWaypoint struct {
	Name string   `xml:"name"`
	Desc string   `xml:"desc"`
	Lat  float64  `xml:"lat,attr"`
	Lon  float64  `xml:"lon,attr"`
	Ele  *float64 `xml:"ele"`
}

func ParseGPXFile(path string) ([]session.GpxTrack, []session.GpxWaypoint, session.Bounds, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, nil, session.Bounds{}, fmt.Errorf("open gpx: %w", err)
	}
	defer f.Close()
	return ParseGPX(f)
}

func ParseGPX(r io.Reader) ([]session.GpxTrack, []session.GpxWaypoint, session.Bounds, error) {
	var doc gpxDoc
	if err := xml.NewDecoder(r).Decode(&doc); err != nil {
		return nil, nil, session.Bounds{}, fmt.Errorf("parse gpx xml: %w", err)
	}

	var tracks []session.GpxTrack
	var waypoints []session.GpxWaypoint

	bounds := session.Bounds{IsSet: false}
	updateBounds := func(lat, lon float64) {
		if !bounds.IsSet {
			bounds = session.Bounds{North: lat, South: lat, East: lon, West: lon, IsSet: true}
			return
		}
		if lat > bounds.North {
			bounds.North = lat
		}
		if lat < bounds.South {
			bounds.South = lat
		}
		if lon > bounds.East {
			bounds.East = lon
		}
		if lon < bounds.West {
			bounds.West = lon
		}
	}

	for _, tr := range doc.Tracks {
		track := session.GpxTrack{Name: tr.Name}
		if track.Name == "" {
			track.Name = "Unnamed Track"
		}
		for _, seg := range tr.Segments {
			for _, pt := range seg.Points {
				ele := 0.0
				if pt.Ele != nil {
					ele = *pt.Ele
				}
				track.Points = append(track.Points, session.GpxPoint{Lat: pt.Lat, Lon: pt.Lon, Elevation: ele})
				updateBounds(pt.Lat, pt.Lon)
			}
		}
		if len(track.Points) >= 2 {
			tracks = append(tracks, track)
		}
	}

	for _, wp := range doc.Waypoints {
		ele := 0.0
		if wp.Ele != nil {
			ele = *wp.Ele
		}
		waypoints = append(waypoints, session.GpxWaypoint{
			Name: wp.Name, Description: wp.Desc,
			Lat: wp.Lat, Lon: wp.Lon, Elevation: ele,
		})
		updateBounds(wp.Lat, wp.Lon)
	}

	return tracks, waypoints, bounds, nil
}
